### PR TITLE
Caching calls to SampleRowKey in CloudBigtableIO.

### PR DESCRIPTION
**Pull request by sduskis**
Also, added a bit of logging and small performance enhancements to CloudBigtableIOIntegrationTest.

### CODE REVIEW

File: bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java [hunk 1]
```diff
@@ -102,44 +116,71 @@ public void testEstimatedAndSplitForSmallTable() throws Exception {
...
+              public void run() {
+                BoundedReader<Result> reader;
+                try {
+                  reader = bundle.createReader(null);
+                  reader.start();
+                  while(reader.getCurrent() != null) {
+                    count.incrementAndGet();
+                    reader.advance();
+                  }
+                  reader.close();
```
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: shouldn't this reader closes in finally block?

**Comment by derjust** [comment]
Can the impact/performance improvement be quantified?
What is the benefit for a table with - let's say - 150GB data with 1 billion rows.

**Comment by sduskis** [comment]
@derjust: This change is going to have minimal impact.  This caches metadata only.

File: bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java [hunk 2]
```diff
@@ -102,44 +116,78 @@ public void testEstimatedAndSplitForSmallTable() throws Exception {
...
-          mutator.mutate(
-              new Put(Bytes.toBytes("row" + i))
-                  .addColumn(COLUMN_FAMILY, QUALIFIER1, getLargeValue()));
+          byte[] largeValue = Bytes.toBytes(RandomStringUtils.randomAlphanumeric(LARGE_VALUE_SIZE));
+          mutator.mutate(new Put(Bytes.toBytes("row" + i)).addColumn(COLUMN_FAMILY, QUALIFIER1,
+            largeValue));
         }
       }
 
+      LOG.info("getSampleKeys() in testEstimatedAndSplitForLargeTable()", rowCount);
```
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: Shouldn't you need "%d" to have rowCount inserted in the message string?

File: bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java [hunk 3]
```diff
@@ -102,44 +116,78 @@ public void testEstimatedAndSplitForSmallTable() throws Exception {
...
       List<SampleRowKeysResponse> sampleRowKeys = getSampleKeys(tableName);
 
       try {
+        LOG.info("Getting Source in testEstimatedAndSplitForLargeTable()", rowCount);
```
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: ditto

File: bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java [hunk 4]
```diff
@@ -102,44 +116,78 @@ public void testEstimatedAndSplitForSmallTable() throws Exception {
...
         BoundedSource<Result> source =
             CloudBigtableIO.read(new CloudBigtableScanConfiguration(projectId, zoneId, clusterId,
                 tableName.getQualifierAsString(), new Scan()));
+        LOG.info("getEstimatedSizeBytes() in testEstimatedAndSplitForLargeTable()", rowCount);
```
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: Get estimated size in bytes in ...

File: bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java [hunk 5]
```diff
@@ -102,44 +116,78 @@ public void testEstimatedAndSplitForSmallTable() throws Exception {
...
+                  reader = bundle.createReader(null);
+                  reader.start();
+                  while(reader.getCurrent() != null) {
+                    count.incrementAndGet();
+                    reader.advance();
+                  }
+                } catch (IOException e) {
+                  LOG.warn("Could not read bundle: %s", e, bundle);
+                } finally {
+                  if (reader != null) {
```
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: nit: BoundedReader is AutoCloseable,  use try () { ... } for readability.

File: bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java [hunk 6]
```diff
@@ -102,44 +116,78 @@ public void testEstimatedAndSplitForSmallTable() throws Exception {
   public void testEstimatedAndSplitForLargeTable() throws Exception {
```
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: The change of this PR is caching SampleRowKeysResponse, can you please verify this change by doing following:

1. getEstimatedSizeBytes()
2.  SampleRowKeys = GetSampleRowKeys();
3. splitIntoBundles()
4. assert same (SampleRowKeys, GetSampleRowKeys())

or have a separate test method that calls GetSampleRowKeys() twice to verify the same SampleRowKeys.

File: bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java [hunk 7]
```diff
@@ -87,13 +97,17 @@ public void testEstimatedAndSplitForSmallTable() throws Exception {
         SampleRowKeysResponse lastSample = sampleRowKeys.get(sampleRowKeys.size() - 1);
         Assert.assertEquals(lastSample.getOffsetBytes(), estimatedSizeBytes);
 
+        LOG.info("Creating Bundles in testEstimatedAndSplitForSmallTable()");
         List<? extends BoundedSource<Result>> bundles =
             source.splitIntoBundles(estimatedSizeBytes / 2 + 1, null);
         // This will be a small table with no splits, so we return HConstants.EMPTY_START_ROW
         // which can't be split.
+        LOG.info("Creatd Bundles in testEstimatedAndSplitForSmallTable()");
```
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: Created

File: bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java (hunk 6 above)
Comment (sduskis) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: I'm going to use the same underlying cached list for the test.

File: bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java (hunk 2 above)
Comment (sduskis) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: The rowCount was extra.. no need to have it here.

File: bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java (hunk 6 above)
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: Can you clarify a bit?
The change in this PA is caching sampleRowKeys, one way to verify that the sampleRowKeys is loaded first and reused later is:

1. SampleRowKeys = GetSampleRowKeys();
2. getEstimatedSizeBytes()
3. splitIntoBundles()
4. assert same (SampleRowKeys, GetSampleRowKeys())

Comment (sduskis) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: 1) SampleRowKeys = source.getSampleRowKeys()
Why do we need 
4) Assert.assertSame(SampleRowKeys, source.getSampleRowKeys())
???

Comment (sduskis) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: Would this help:

Assert.assertSame(source.getSampleRowKeys(), source.getSampleRowKeys())

Comment (sduskis) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: Oh... I see.  I only updated one test.  I'll update the other test as well.

File: bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java [hunk 8]
```diff
@@ -76,23 +84,31 @@ public void testEstimatedAndSplitForSmallTable() throws Exception {
...
         Assert.assertEquals(lastSample.getOffsetBytes(), estimatedSizeBytes);
 
+        LOG.info("Creating Bundles in testEstimatedAndSplitForSmallTable()");
         List<? extends BoundedSource<Result>> bundles =
             source.splitIntoBundles(estimatedSizeBytes / 2 + 1, null);
         // This will be a small table with no splits, so we return HConstants.EMPTY_START_ROW
         // which can't be split.
+        LOG.info("Created Bundles in testEstimatedAndSplitForSmallTable()");
         Assert.assertEquals(sampleRowKeys.size() * 2 - 1, bundles.size());
+        Assert.assertSame(source, source.getSampleRowKeys());
```
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java]: (sampleRowKeys, ...)

**Comment by kevinsi4508** [comment]
LGTM

**Pull request merged**